lmdeploy/serve/async_engine.py

```diff
@@ -118,21 +136,26 @@ async def generate(
             repetition_penalty (float): The parameter for repetition penalty.
               1.0 means no penalty
             ignore_eos (bool): indicator for ignoring eos
+            do_preprocess (bool): whether pre-process the messages.
         """
         instance_id = session_id % self.instance_num
         if str(session_id) not in self.steps:
             self.steps[str(session_id)] = 0
         if step != 0:
             self.steps[str(session_id)] = step
         seed = random.getrandbits(64)
-        prompt = self.model.messages2prompt(messages, sequence_start)
+        prompt = messages
+        if do_preprocess:
+            prompt = self.model.messages2prompt(prompt, sequence_start)
         input_ids = self.tokenizer.encode(prompt)
         finish_reason = 'stop' if stop else None
         if self.steps[str(session_id)] + len(
```

Comment:
I think `+ self.request_output_len` is missing

Reply:
We found the model can still output something shorter than `max_tokens`.
https://github.com/InternLM/lmdeploy/blob/56942c43315af2a205efbaaa6343786f6e11c943/lmdeploy/turbomind/chat.py#L122-L125